ksqldb-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java

```diff
@@ -1341,6 +1341,18 @@ private static ConfigDef buildConfigDef(final ConfigGeneration generation) {
   }
   // CHECKSTYLE_RULES.ON: MethodLength
 
+  public Map<String, Object> originalsWithPrefixOverride(final String prefix) {
+    final Map<String, Object> originals = originals();
+    final Map<String, Object> result = new HashMap<>(originals);
+    for (Map.Entry<String, ?> entry : originals.entrySet()) {
+      if (entry.getKey().startsWith(prefix) && entry.getKey().length() > prefix.length()) {
+        final String keyWithNoPrefix = entry.getKey().substring(prefix.length());
+        result.put(keyWithNoPrefix, entry.getValue());
```

Comment:
Nit: should we also `result.delete(entry.getKey())` to avoid a double entry? In the end, Kafka Streams won't understand the prefixed key, and IIRC would log a warning about an unknown config?

Reply:
What if instead calling delete, you avoid populating the hashmap during initialization, and just add all of them in the for loop, but removing the prefix for those with the prefix found?. It'll save time on adding on initialization, then deleting on the loop.

i.e.
```
# initialize the initial capacity to save time on resizing internally
final Map<String, Object> result = new HashMap<>(originals.size());

for (Map.Entry<String, ?> entry : originals.entrySet()) {
   if (prefixed) {
     ... add with prefix removed
   } else {
     ... add not prefixed value
   }
}

```

Reply:
Nevermind. I understand the code now. Originals are added first, and then not prefixed values are overriden by a prefixed value. 

Regarding the question if calling `delete` to avoid duplication. If the duplicated value will be ignored later anyway, then we can save time by not calling delete here, right?

Reply:
I don't think it would make much of a difference from a perf point of view if we delete them or not. -- But if we can avoid unnecessary WARN logs, it might be better to delete them?

Reply:
It absolutely makes sense. I was even considering it in the beginning. I was wondering whether we should retain or not the prefixed values.
This logic anyway should be in the base config classes.

Reply:
> This logic anyway should be in the base config classes.

What logic are you referring, too? To drop unknown configs? It would not be correct to do it in the base config class, because we have an architecture of nesting -- thus, it must be possible to add any custom config and it's expected that it's just forwarded. (We still log a WARN for this case to guard against typos.)

Reply:
I mean the logic to get the originals with prefix overrides.

We have already the logic for values:
https://kafka.apache.org/30/javadoc/org/apache/kafka/common/config/AbstractConfig.html#originalsWithPrefix(java.lang.String,boolean)

This, however, includes filtering of the entries to the available keys only. The code I added is a counterpart where the configuration keys cannot be determined at the query time.